### PR TITLE
Userid revocations and designated key revokers

### PR DESCRIPTION
**Pull request by zapu**
work in progress

Revoked identities won't show up in `Identities` map, but they can be accessed through `BadIdentities`. Also there is a `Revocation` field in `Identity` struct now.

Designated revokes end up in `UnverifiedRevocations`. Only revocations issued by designated keys are considered. Designation has to be signed and this is verified, but the consumer has to verify actual revocation (using the designee's key).

### CODE REVIEW

File: openpgp/read.go [hunk 1]
```diff
@@ -461,3 +461,31 @@ func checkArmoredDetachedSignature(keyring KeyRing, signed, signature io.Reader)
 	}
 	return checkDetachedSignature(keyring, signed, body)
 }
+
+// CheckDesignatedRevokers will try to confirm any of designated
+// revocation of entity. For this function to work, revocation
+// issuer's key should be found in keyring, otherwise it will return
+// false, nil if keys are not found (meaning: revocation not
+// confirmed).
+func CheckDesignatedRevokers(keyring KeyRing, entity *Entity) (bool, *Key) {
```
Comment (zapu) [inline openpgp/read.go]: Still unsure about the API here. Entity can be basically in 3 states now, in regards to revocations:
1) Not revoked
2) Maybe revoked (if there are 3rd party revocations we couldn't verify)
3) Definitely revoked (if there are 1st or 3rd party revocations that were verified)

File: openpgp/packet/public_key.go [hunk 2]
```diff
@@ -598,7 +598,7 @@ func (pk *PublicKey) CanSign() bool {
 	return pk.PubKeyAlgo != PubKeyAlgoRSAEncryptOnly && pk.PubKeyAlgo != PubKeyAlgoElGamal
 }
 
-// VerifySignature returns nil iff sig is a valid signature, made by this
+// VerifySignature returns nil if sig is a valid signature, made by this
```
Comment (maxtaco) [inline openpgp/packet/public_key.go]: why these changes? `iff` == "if and only if".  I think it's intentional.

File: openpgp/read.go (hunk 1 above)
Comment (maxtaco) [reply openpgp/read.go]: Where is this called from?

**Review by maxtaco — COMMENTED** [review]
Some comments / questions.

File: openpgp/packet/public_key.go (hunk 2 above)
Comment (zapu) [reply openpgp/packet/public_key.go]: I had no idea. I'll change these back. I run my editor with spell-checker and I occasionally correct a comment if it's near code I'm working on.

File: openpgp/read.go (hunk 1 above)
Comment (zapu) [reply openpgp/read.go]: Right now I'm calling this from client. So for example, `CheckDetachedSignature` would return a `signer` and then client would call `CheckDesignatedRevokers` with that signer. I'm not sure if it should be called automatically from e.g. `ReadMessage`.

File: openpgp/keys.go [hunk 3]
```diff
@@ -500,13 +520,26 @@ EachPacket:
...
-			} else if pkt.SigType == packet.SigTypeKeyRevocation {
-				// These revocations won't revoke UIDs as handled above, so lookout!
-				revocations = append(revocations, pkt)
+			} else if current != nil && pkt.SigType == packet.SigTypeIdentityRevocation {
+				if err = e.PrimaryKey.VerifyUserIdSignature(current.Name, e.PrimaryKey, pkt); err == nil {
+					current.Revocation = pkt
+
+					// Remove current from Identities, because the revocation signature
+					// might come after the certification signature.
+					delete(e.Identities, current.Name)
```
Comment (zapu) [inline openpgp/keys.go]: but few lines before is:
```
// NOTE! We might later see a revocation for this very same UID, and it
// won't be undone. We've preserved this feature from the original
// Google OpenPGP we forked from.
```

what should we do?

Comment (maxtaco) [reply openpgp/keys.go]: If we make PGP more restrictive (i.e., disallow keys we previously allowed), we might be creating a ton of ~~bugs~~ support requests for ourselves down the road....

Comment (maxtaco) [reply openpgp/keys.go]: The other thing is that we're not resetting `current.SelfSignature` in this case, which might be required.

File: openpgp/read.go (hunk 1 above)
Comment (maxtaco) [reply openpgp/read.go]: I see

File: openpgp/keys.go (hunk 3 above)
Comment (maxtaco) [reply openpgp/keys.go]: I'm OK to give this a try, but we have to be willing to back it out.

Comment (zapu) [reply openpgp/keys.go]: Good point about `SelfSignature`. But for some keys it would have to go back and find another `SelfSignature`, from other uid.. So maybe lets not invalidate any uids, keep everything in `Identities` list but have the `Revocation` field in Identity is some code wanted to filter out revoked ones?

Comment (maxtaco) [reply openpgp/keys.go]: Meaning get rid of `BadIdentities`?

Comment (zapu) [reply openpgp/keys.go]: Yes. Otherwise whether revoked uid lands in Identities or BadIdentities is decided by order of signatures. If revocation comes first, it'll go to BadIdentities, if certification comes first, it'll go to Identities (but both cases will set `current.Revocation` field).

Comment (maxtaco) [reply openpgp/keys.go]: And just to double-check, we previously weren't checking these revocations at all?

Comment (zapu) [reply openpgp/keys.go]: Correct.

Comment (maxtaco) [reply openpgp/keys.go]: ok, i'm on board.  nix the `BadIdentities`, sorry it was the wrong idea.

**Review by maxtaco — APPROVED** [review]

File: openpgp/keys.go [hunk 4]
```diff
@@ -442,6 +452,13 @@ EachPacket:
 			current.Name = pkt.Id
 			current.UserId = pkt
 		case *packet.Signature:
+			if pkt.SigType == packet.SigTypeKeyRevocation {
+				// These revocations won't revoke UIDs (see
+				// SigTypeIdentityRevocation). Handle these first,
+				// because key might have revocation coming from
+				// another key (designated revoke).
+				revocations = append(revocations, pkt)
```
Comment (maxtaco) [inline openpgp/keys.go:460]: What are implications of now allowing this case to fallthrough below?

Comment (maxtaco) [reply openpgp/keys.go:460]: I.e., we're also appending this onto `current.Signatures`, whereas previous we weren't? Do I read that properly and is that OK?

Comment (zapu) [reply openpgp/keys.go:460]: This is moved from line 503, i.e. it used to only append to `revocations` when `pkt.IssuerKeyId != nil && *pkt.IssuerKeyId != e.PrimaryKey.KeyId` did not evaluate to true. Now it always appends to `revocations` but that check is now at line 581 (for 1st party revocation) and there is additional logic regarding 3rd party revocations.

Comment (maxtaco) [reply openpgp/keys.go:460]: That's not my question.  My question is that now these packets are going to hit line 551 in the new code, whereas previous it did not (since it was `if .. else` gated and handled by a different condition).  So all of these packets will be in the `current.Signatures` array as a result of line 551.  Does that matter?

Comment (zapu) [reply openpgp/keys.go:460]: You're right, I missed that. It should not go to `current.Signatures`.

Comment (maxtaco) [reply openpgp/keys.go:460]: ok, so if it's not intentional, let's fix it...

Comment (zapu) [reply openpgp/keys.go:460]: Fixed one line below that (`continue`)

**Review by maxtaco — APPROVED** [review]
